src/PantherActions.php

```diff
@@ -278,49 +272,53 @@ protected static function submitForm(string $selector): void
 
     protected static function findFormField(string $fieldText, string $legend = null, string $contextSelector = null): Crawler
     {
-        $xpath = <<<'XPATH'
-            [
-                contains(concat(' ', normalize-space(string(.)), ' '), %1$s) or
-                contains(concat(' ', normalize-space(string(@value)), ' '), %1$s) or
-                @id=%2$s or
-                @name=%2$s or
-                @placeholder=%2$s
-            ]
-            XPATH;
-
+        // create crawler, optionally start by context
         $crawler = self::crawlerBySelector($contextSelector);
 
+        // if legend is given, search for that and get its fieldset
         if ($legend !== null) {
-            $crawler = $crawler
-                ->filterXPath(
-                    sprintf(
-                        '//legend' . $xpath . '/ancestor::fieldset',
-                        Crawler::xpathLiteral(' ' . $legend . ' '),
-                        Crawler::xpathLiteral($legend),
-                    )
-                )
-            ;
+            $crawler = $crawler->filterXPath(
+                self::formFieldXpath('descendant-or-self::legend', $legend) . '/ancestor::fieldset'
```

Comment:
Different from what we expected, the `//` always searches in the entire document, even when it's already filtered. What we were looking for was `descendant-or-self::`. I've fixed that here and added a test for it.